modules/phase_field/src/materials/AsymmetricCrossTermBarrierFunctionMaterial.C

```diff
@@ -0,0 +1,97 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "AsymmetricCrossTermBarrierFunctionMaterial.h"
+
+template<>
+InputParameters validParams<AsymmetricCrossTermBarrierFunctionMaterial>()
+{
+  InputParameters params = validParams<CrossTermBarrierFunctionBase>();
+  params.addClassDescription("Free energy contribution asymmetric across interfaces between arbitrary pairs of phases.");
+  params.addParam<std::vector<MaterialPropertyName> >("hi_names", "Switching Function Materials that provide h(eta_i)");
+  return params;
+}
+
+AsymmetricCrossTermBarrierFunctionMaterial::AsymmetricCrossTermBarrierFunctionMaterial(const InputParameters & parameters) :
+    CrossTermBarrierFunctionBase(parameters),
+    _h(_num_eta),
+    _dh(_num_eta),
+    _d2h(_num_eta)
+{
+  // switching functions
+  const std::vector<MaterialPropertyName> & hi_names = getParam<std::vector<MaterialPropertyName> >("hi_names");
+  if (hi_names.size() != _num_eta)
+    mooseError("The number of coupled etas must be equal to the number of hi_names in AsymmetricCrossTermBarrierFunctionMaterial " << name());
+
+  for (unsigned int i = 0; i < _num_eta; ++i)
+  {
+    _h[i] = &getMaterialProperty<Real>(hi_names[i]);
+    _dh[i] = &getMaterialPropertyDerivative<Real>(hi_names[i], _eta_names[i]);
+    _d2h[i] = &getMaterialPropertyDerivative<Real>(hi_names[i], _eta_names[i], _eta_names[i]);
+  }
+}
+
+void
+AsymmetricCrossTermBarrierFunctionMaterial::computeQpProperties()
+{
+  // Initialize properties to zero before accumulating
+  CrossTermBarrierFunctionBase::computeQpProperties();
+
+  // Sum the components of our W_ij matrix to get constant used in our g function
+  for (unsigned int i = 0; i < _num_eta; ++i)
+    for (unsigned int j = i + 1; j < _num_eta; ++j)
+    {
+      const Real ni = (*_eta[i])[_qp];
+      const Real nj = (*_eta[j])[_qp];
+
+      const Real Wij = _W_ij[_num_eta * i + j];
+      const Real Wji = _W_ij[_num_eta * j + i];
+
```

Comment:
What you've done with `Wij`, `Wji`, `ni` and `nj` looks very clean, want to do the same for `_h[i]`, `_dh[i]`, etc?